src/crypto/CHIPCryptoPAL.cpp

```diff
@@ -922,6 +922,91 @@ CHIP_ERROR DeriveGroupOperationalCredentials(const ByteSpan & epoch_key, const B
+                                              uint16_t vendorId, MutableByteSpan & outputSpan)
+{
+    // Only V1 supported yet.
+    VerifyOrReturnError(fabricBindingVersion == kFabricBindingVersionV1, CHIP_ERROR_INVALID_ARGUMENT);
+
+    VerifyOrReturnError(outputSpan.size() >= kVendorFabricBindingMessageV1MaxSize, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    Encoding::BigEndian::BufferWriter writer(outputSpan.data(), outputSpan.size());
+
+    // vendor_fabric_binding_message := fabric_binding_version (1 byte) || root_public_key || fabric_id || vendor_id
+    writer.Put8(fabricBindingVersion).Put(rootPublicKey.ConstBytes(), rootPublicKey.Length()).Put64(fabricId).Put16(vendorId);
+
+    size_t actuallyWritten = 0;
+    if (!writer.Fit(actuallyWritten))
+    {
+        return CHIP_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    outputSpan.reduce_size(actuallyWritten);
+    return CHIP_NO_ERROR;
+}
+
+/**
+ * @brief Generate the message to be signed for the Fabric Table Vendor ID Verification Procedure.
+ *
+ * The Fabric Binding Version value will be recovered from the vendorFabricBindingMessage.
+ *
+ * @param fabricIndex - Fabric Index for the fabric in question
+ * @param clientChallenge - Client challenge to use
+ * @param attestationChallenge - Attestation challenge to use
+ * @param vendorFabricBindingMessage - The VendorFabricBindingMessage previously computed for the fabric
+ * @param vidVerificationStatement - The VID Verification Statement to include in signature (may be empty)
+ * @param outputSpan - Span that will receive the to-be-signed message. Must be large enough for the
+ *                     payload (otherwise CHIP_ERROR_BUFFER_TOO_SMALL) and will be resized to fit.
+ * @return CHIP_NO_ERROR on success, otherwise another CHIP_ERROR value representative of the failure.
+ */
+CHIP_ERROR GenerateVendorIdVerificationToBeSigned(FabricIndex fabricIndex, const ByteSpan & clientChallenge,
+                                                  const ByteSpan & attestationChallenge,
+                                                  const ByteSpan & vendorFabricBindingMessage,
+                                                  const ByteSpan & vidVerificationStatement, MutableByteSpan & outputSpan)
+{
+    VerifyOrReturnError((clientChallenge.size() == kVendorIdVerificationClientChallengeSize) &&
+                            (attestationChallenge.size() == CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES) &&
+                            !vendorFabricBindingMessage.empty(),
+                        CHIP_ERROR_INVALID_ARGUMENT);
+
+    // Extract binding version from vendorFabricBindingMessage. Only V1 supported yet.
+    uint8_t fabricBindingVersion = vendorFabricBindingMessage[0];
+    VerifyOrReturnError(fabricBindingVersion == kFabricBindingVersionV1, CHIP_ERROR_INVALID_ARGUMENT);
+
+    VerifyOrReturnError(outputSpan.size() >= kVendorIdVerificationTbsV1MaxSize, CHIP_ERROR_BUFFER_TOO_SMALL);
```

Comment:
do we need this? Would `fit` not already catch this?

Reply:
I think that would work. I was trying to avoid touching buffer at all if too small.